### PR TITLE
[deckhouse] release webhook notification

### PR DESCRIPTION
**Pull request by yalosev**
## Description
Add two parameters for deckhouse settings:
`deckhouse.update.notification.webhook` - URL of a webhook, where JSON with release details will be send
`deckhouse.update.notification.minimalNotificationTime` - the minimum amount of time we have to warn the user about a release 

## Why do we need it, and what problem does it solve?
Add an ability to control and plan deckhouse releases for users

## Checklist
- [x] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: deckhouse
type: feature
summary: Add the ability to notify about the upcoming Deckhouse release via a webhook.
impact_level: default
```

### CODE REVIEW

File: modules/020-deckhouse/docs/USAGE_RU.md [hunk 1]
```diff
@@ -94,6 +94,54 @@ deckhouse: |
 kubectl annotate DeckhouseRelease v1-36-0 release.deckhouse.io/disruption-approved=true
 ```
 
+### Оповещение об обновлении Deckhouse
+
+В режиме обновлений `Auto` можно настроить оповещение о предстоящих релизах deckhouse, например:
```
Comment (z9r5) [inline modules/020-deckhouse/docs/USAGE_RU.md:99]:
```suggestion
В режиме обновлений `Auto` можно настроить оповещение о предстоящих релизах Deckhouse, например:
```

File: modules/020-deckhouse/docs/USAGE_RU.md [hunk 2]
```diff
@@ -94,6 +94,54 @@ deckhouse: |
...
+```yaml
+deckhouse: |
+  ...
+  update:
+    mode: Auto
+    notification:
+      webhook: https://release-webhook.mydomain.com
+```
+
+При таких настройках, на адрес `https://release-webhook.mydomain.com` придет POST запрос с `Content-Type: application/json` и следующим содержианием:
```
Comment (z9r5) [inline modules/020-deckhouse/docs/USAGE_RU.md:110]:
```suggestion
При таких настройках, на адрес `https://release-webhook.mydomain.com` придет POST-запрос с `Content-Type: application/json` и следующим содержанием:
```

File: modules/020-deckhouse/docs/USAGE_RU.md [hunk 3]
```diff
@@ -94,6 +94,54 @@ deckhouse: |
...
+```json
+{
+  "version": "1.36",
+  "requirements":  {"k8s": "1.20.0"}, 
+  "changelogLink": "https://github.com/deckhouse/deckhouse/changelog/1.36.md",
+  "applyTime": "2023-01-01T14:30:00Z00:00",
+  "message": "New Deckhouse Release 1.36 is available. Release will be applied at: Friday, 01-Jan-22 14:30:00 UTC"
+}
+```
+
```
Comment (z9r5) [inline modules/020-deckhouse/docs/USAGE_RU.md:121]: Описание полей:

Comment (z9r5) [inline modules/020-deckhouse/docs/USAGE_RU.md:121]:
```suggestion

Описание полей:
```

File: modules/020-deckhouse/docs/USAGE_RU.md [hunk 4]
```diff
@@ -94,6 +94,54 @@ deckhouse: |
...
+- `version` - строка, версия нового релиза
+- `requirements` - объект, требования к новому релизу
+- `changelogLink` - строка, ссылка на документацию Changelog нового релиза
+- `applyTime` - строка, дата и время обновления (расчитывается по готовности релиза и вашим окнам обновлений). Формат: RFC3339
+- `message` - строка, текстовое сообщение о доступности и времени обновления
```
Comment (z9r5) [inline modules/020-deckhouse/docs/USAGE_RU.md:126]:
```suggestion
- `version` — строка, версия нового релиза;
- `requirements` — объект, требования к новому релизу;
- `changelogLink` — строка, ссылка на список изменений (changelog) нового релиза;
- `applyTime` — строка, дата и запланированное время обновления (расчитывается по готовности релиза и установленным окнам обновлений) в формате `RFC3339`;
- `message` — строка, текстовое сообщение о доступности нового релиза и запланированного времени обновления.
```

File: modules/020-deckhouse/docs/USAGE_RU.md [hunk 5]
```diff
@@ -94,6 +94,54 @@ deckhouse: |
...
+  ...
+  update:
+    mode: Auto
+    notification:
+      webhook: https://release-webhook.mydomain.com
+      minimalNotificationTime: 8h
+```
+
+и задает минимальное время до обновления, за которое должно придти оповещение. Механизм обновления Deckhouse гарантирует,
+что Deckhouse не обновится раньше указанного время. Если вы используете окна обновлений, то обновление Deckhouse произойдет в следующее возможное окно обновлений.
```
Comment (z9r5) [inline modules/020-deckhouse/docs/USAGE_RU.md:143]:
```suggestion

Задает минимальное время которое должно пройти перед обновлением с момента оповещения. Механизм обновления Deckhouse гарантирует, что Deckhouse не обновится раньше указанного времени. Если вы используете окна обновлений, то обновление Deckhouse произойдет в ближайшее возможное окно обновлений.

Пример:

```yaml
deckhouse: |
  ...
  update:
    mode: Auto
    notification:
      webhook: https://release-webhook.mydomain.com
      minimalNotificationTime: 8h
```
```

**Review by z9r5 — APPROVED** [review]
Made some changes

File: modules/020-deckhouse/docs/USAGE_RU.md [hunk 6]
```diff
@@ -94,6 +94,53 @@ deckhouse: |
...
+  "applyTime": "2023-01-01T14:30:00Z00:00",
+  "message": "New Deckhouse Release 1.36 is available. Release will be applied at: Friday, 01-Jan-22 14:30:00 UTC"
+}
+```
+
+Чтобы всегда иметь достаточно времени для реакции на оповещение об обновлении Deckhouse, достаточно настроить параметр [minimalNotificationTime](configuration.html#parameters-update-notification-minimalnotificationtime). В этом случае обновление случится по прошествии указанного времени, с учетом окон обновлений.
+
+#### Минимальное время оповещения об обновлении
+
+
```
Comment (z9r5) [inline modules/020-deckhouse/docs/USAGE_RU.md:130]:
```suggestion

```

File: modules/020-deckhouse/docs/USAGE.md [hunk 7]
```diff
@@ -94,6 +94,49 @@ In this mode, it will be necessary to confirm each minor disruptive update with
 kubectl annotate DeckhouseRelease v1-36-0 release.deckhouse.io/disruption-approved=true
 ```
 
+### Deckhouse update notification
+
+In the `Auto` update mode, you can [set up](configuration.html#parameters-update-notification) a webhook call, to be notified of an upcoming Deckhouse minor version update. 
+
```
Comment (z9r5) [inline modules/020-deckhouse/docs/USAGE.md:100]:
```suggestion
In the `Auto` update mode, you can [set up](configuration.html#parameters-update-notification) a webhook call, to be notified of an upcoming Deckhouse minor version update.

```

File: modules/020-deckhouse/docs/USAGE_RU.md [hunk 8]
```diff
@@ -94,6 +94,51 @@ deckhouse: |
...
+  "requirements":  {"k8s": "1.20.0"}, 
+  "changelogLink": "https://github.com/deckhouse/deckhouse/changelog/1.36.md",
+  "applyTime": "2023-01-01T14:30:00Z00:00",
+  "message": "New Deckhouse Release 1.36 is available. Release will be applied at: Friday, 01-Jan-22 14:30:00 UTC"
+}
+```
+
+Чтобы всегда иметь достаточно времени для реакции на оповещение об обновлении Deckhouse, достаточно настроить параметр [minimalNotificationTime](configuration.html#parameters-update-notification-minimalnotificationtime). В этом случае обновление случится по прошествии указанного времени, с учетом окон обновлений.
+
+#### Минимальное время оповещения об обновлении
```
Comment (z9r5) [inline modules/020-deckhouse/docs/USAGE_RU.md:128]:
```suggestion

```

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/test'...

File: modules/020-deckhouse/hooks/internal/updater/notification.go [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+	return &NotificationConfig{
+		WebhookURL:              webhook.String(),
+		MinimalNotificationTime: minimalTime,
+	}
+}
+
+func sendWebhookNotification(webhookURL string, data webhookData) error {
+	tr := &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
```
Comment (name212) [inline modules/020-deckhouse/hooks/internal/updater/notification.go:59]: In another products, I always see that this `InsecureSkipVerify` are passing with URL as separate parameter. Client could choice skipping verify. Or we should say about skip TLS check in [documentation](https://github.com/deckhouse/deckhouse/pull/2131/files#diff-6062960603271f16b22204e3e46807bc9333f9ad4d68adc866826e17064be2e5R112), i think.

File: modules/020-deckhouse/hooks/internal/updater/notification.go [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+	}
+	client := &http.Client{
+		Transport: tr,
+		Timeout:   10 * time.Second,
+	}
+
+	buf := bytes.NewBuffer(nil)
+	_ = json.NewEncoder(buf).Encode(data)
+
+	_, err := client.Post(webhookURL, "application/json", buf)
```
Comment (name212) [inline modules/020-deckhouse/hooks/internal/updater/notification.go:69]: Would we add back off for 3 time if request was failed?

File: modules/020-deckhouse/hooks/internal/updater/notification.go (hunk 9 above)
Comment (yalosev) [reply modules/020-deckhouse/hooks/internal/updater/notification.go]: Separated with config

File: modules/020-deckhouse/hooks/internal/updater/notification.go (hunk 10 above)
Comment (yalosev) [reply modules/020-deckhouse/hooks/internal/updater/notification.go]: makes sense

File: modules/020-deckhouse/hooks/internal/updater/updater.go [hunk 11]
```diff
@@ -0,0 +1,654 @@
...
+	if du.inManualMode || du.releaseData.Notified {
+		return true
+	}
+
+	var applyTimeChanged bool
+	predictedReleaseApplyTime := du.predictedReleaseApplyTime(predictedRelease)
+	if du.notificationConfig.MinimalNotificationTime.Duration > 0 {
+		minApplyTime := du.now.Add(du.notificationConfig.MinimalNotificationTime.Duration)
+		if minApplyTime.Before(predictedReleaseApplyTime) {
+			minApplyTime = predictedReleaseApplyTime
```
Comment (name212) [inline modules/020-deckhouse/hooks/internal/updater/updater.go:116]: Can you add (or show me) test for this code. I broke my mind when i read this

**Review by name212 — APPROVED** [review]

File: modules/020-deckhouse/hooks/internal/updater/updater.go (hunk 11 above)
Comment (yalosev) [reply modules/020-deckhouse/hooks/internal/updater/updater.go:116]: "Notification: release applyAfter time is after notification period" test is about that

**Review by nabokihms — APPROVED** [review]

**Pull request merged**